lib/rubocop/cop/lint/format_parameter_mismatch.rb

```diff
@@ -0,0 +1,92 @@
+# encoding: utf-8
+
+module RuboCop
+  module Cop
+    module Lint
+      # This lint sees if there is a mismatch between the number of
+      # expected fields for format/sprintf/#% and what is actually
+      # passed as arguments.
+      #
+      # @example
+      #
+      #   format('A value: %s and another: %i', a_value)
+      #
+      class FormatParameterMismatch < Cop
+        MSG = 'Number of fields (%i) mismatches number of arguments (%i).'
+        VALID_FIELDS_REGEX = /%[bBdiouxXeEfgGacps]/
+
+        def fields_regex
+          VALID_FIELDS_REGEX
+        end
+
+        def on_send(node)
+          add_offense(node, :selector) if offending_node?(node)
+        end
+
+        private
+
+        def offending_node?(node)
+          if sprintf?(node) || format?(node) || percent?(node)
+            num_of_args_for_format, num_of_expected_fields = count_matches(node)
+            num_of_expected_fields != num_of_args_for_format
+          else
+            false
+          end
+        end
+
+        def count_matches(node)
+          receiver_node, _method_name, *args = *node
+
+          if sprintf?(node) || format?(node)
+            number_of_args_for_format = (args.size - 1)
+            number_of_expected_fields = expected_fields(args.first).size
+          elsif percent?(node)
+            number_of_args_for_format = args.first.child_nodes.size
+            number_of_expected_fields = expected_fields(receiver_node).size
+          end
+
+          [number_of_args_for_format, number_of_expected_fields]
+        end
+
+        def format_method?(name, node)
+          receiver, method_name, *args = *node
+
+          # commands have no explicit receiver
+          return false unless !receiver && method_name == name
+
+          args.size > 1 && args.first.type != :const
+        end
+
+        def expected_fields(node)
```

Comment:
I think it'd be better if this method operated just on a string. This will make it easier to comprehend. Drilling to the source is kind of distracting.